File: by-language/python/algorithms/sorting.py
"""
Sorting Algorithms

╔════════════════════════════════════════════════════════════════════════════╗
║                         SORTING ALGORITHMS OVERVIEW                        ║
╠════════════════════════════════════════════════════════════════════════════╣
║                                                                            ║
║  Comparison-Based Sorts:                                                   ║
║  ┌─────────────────┬────────────┬────────────┬────────────┬────────┐       ║
║  │ Algorithm       │ Best       │ Average    │ Worst      │ Stable │       ║
║  ├─────────────────┼────────────┼────────────┼────────────┼────────┤       ║
║  │ Bubble Sort     │ O(n)       │ O(n²)      │ O(n²)      │ Yes    │       ║
║  │ Selection Sort  │ O(n²)      │ O(n²)      │ O(n²)      │ No     │       ║
║  │ Insertion Sort  │ O(n)       │ O(n²)      │ O(n²)      │ Yes    │       ║
║  │ Merge Sort      │ O(n log n) │ O(n log n) │ O(n log n) │ Yes    │       ║
║  │ Quick Sort      │ O(n log n) │ O(n log n) │ O(n²)      │ No     │       ║
║  │ Heap Sort       │ O(n log n) │ O(n log n) │ O(n log n) │ No     │       ║
║  │ Shell Sort      │ O(n log n) │ O(n^1.3)   │ O(n²)      │ No     │       ║
║  └─────────────────┴────────────┴────────────┴────────────┴────────┘       ║
║                                                                            ║
║  Non-Comparison Sorts:                                                     ║
║  ┌─────────────────┬────────────┬────────────┬────────────┬────────┐       ║
║  │ Algorithm       │ Time       │ Space      │ Constraint │ Stable │       ║
║  ├─────────────────┼────────────┼────────────┼────────────┼────────┤       ║
║  │ Counting Sort   │ O(n + k)   │ O(k)       │ Integers   │ Yes    │       ║
║  │ Radix Sort      │ O(d(n+k))  │ O(n + k)   │ Integers   │ Yes    │       ║
║  │ Bucket Sort     │ O(n + k)   │ O(n)       │ Uniform    │ Yes    │       ║
║  └─────────────────┴────────────┴────────────┴────────────┴────────┘       ║
║                                                                            ║
║  Bubble Sort Visualization:                                                ║
║  [5, 3, 8, 4, 2] → [3, 5, 4, 2, 8] → [3, 4, 2, 5, 8] → ... → [2,3,4,5,8]   ║
║    ↑ ↑ compare & swap if needed, largest bubbles to end                    ║
║                                                                            ║
║  Merge Sort Visualization:                                                 ║
║  [5, 3, 8, 4, 2, 7, 1, 6]                                                  ║
║        /               \                  Split                            ║
║  [5, 3, 8, 4]    [2, 7, 1, 6]                                              ║
║      /    \         /    \               Split more                        ║
║  [5,3]  [8,4]   [2,7]  [1,6]                                               ║
║    |      |       |      |               Merge pairs                       ║
║  [3,5]  [4,8]   [2,7]  [1,6]                                               ║
║      \    /         \    /               Merge                             ║
║  [3, 4, 5, 8]    [1, 2, 6, 7]                                              ║
║        \               /                 Final merge                       ║
║  [1, 2, 3, 4, 5, 6, 7, 8]                                                  ║
║                                                                            ║
║  Quick Sort (Partition):                                                   ║
║  [5, 3, 8, 4, 2, 7, 1, 6]  pivot=5                                         ║
║  [3, 4, 2, 1] [5] [8, 7, 6]  elements < 5 | pivot | elements > 5           ║
║                                                                            ║
╚════════════════════════════════════════════════════════════════════════════╝

LEETCODE PROBLEMS:
- #912 Sort an Array
- #75 Sort Colors (Dutch National Flag)
- #148 Sort List (Merge Sort on Linked List)
- #215 Kth Largest Element (Quick Select)
- #347 Top K Frequent Elements
- #56 Merge Intervals (sort then merge)
- #179 Largest Number (custom comparator)
- #973 K Closest Points to Origin

USE CASES:
- Bubble/Selection/Insertion: Small arrays, nearly sorted data, educational
- Merge Sort: Stable sort needed, linked lists, external sorting
- Quick Sort: General purpose, in-place sorting, cache-friendly
- Heap Sort: Guaranteed O(n log n), memory-constrained
- Counting/Radix: Integer sorting, known range, linear time needed
- Bucket Sort: Uniformly distributed floating-point numbers
"""

from typing import TypeVar, List, Callable, Optional
import random

T = TypeVar('T')


# =============================================================================
# COMPARISON-BASED SORTING ALGORITHMS
# =============================================================================


def bubble_sort(arr: List[T]) -> List[T]:
    """
    Bubble Sort - repeatedly swap adjacent elements if in wrong order.

    Time Complexity: O(n²) average/worst, O(n) best (already sorted)
    Space Complexity: O(1)
    Stable: Yes

    Args:
        arr: List to sort.

    Returns:
        New sorted list.
    """
    result = arr.copy()
    n = len(result)

    for i in range(n):
        swapped = False

        for j in range(0, n - i - 1):
            if result[j] > result[j + 1]:
                result[j], result[j + 1] = result[j + 1], result[j]
                swapped = True

        # Optimization: if no swaps, array is sorted
        if not swapped:
            break

    return result


def selection_sort(arr: List[T]) -> List[T]:
    """
    Selection Sort - find minimum and place at beginning.

    Time Complexity: O(n²) all cases
    Space Complexity: O(1)
    Stable: No (can be made stable with insertion instead of swap)

    Args:
        arr: List to sort.

    Returns:
        New sorted list.
    """
    result = arr.copy()
    n = len(result)

    for i in range(n):
        min_idx = i

        for j in range(i + 1, n):
            if result[j] < result[min_idx]:
                min_idx = j

        result[i], result[min_idx] = result[min_idx], result[i]

    return result


def insertion_sort(arr: List[T]) -> List[T]:
    """
    Insertion Sort - build sorted portion one element at a time.

    Time Complexity: O(n²) average/worst, O(n) best (nearly sorted)
    Space Complexity: O(1)
    Stable: Yes

    Excellent for small arrays and nearly sorted data.
    Used as base case in hybrid sorts (Timsort, Introsort).

    Args:
        arr: List to sort.

    Returns:
        New sorted list.
    """
    result = arr.copy()
    n = len(result)

    for i in range(1, n):
        key = result[i]
        j = i - 1

        while j >= 0 and result[j] > key:
            result[j + 1] = result[j]
            j -= 1

        result[j + 1] = key

    return result


def merge_sort(arr: List[T]) -> List[T]:
    """
    Merge Sort - divide and conquer with merging.

    Time Complexity: O(n log n) all cases
    Space Complexity: O(n)
    Stable: Yes

    Preferred for linked lists and when stability is required.

    Args:
        arr: List to sort.

    Returns:
        New sorted list.
    """
    if len(arr) <= 1:
        return arr.copy()

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return _merge(left, right)


def _merge(left: List[T], right: List[T]) -> List[T]:
    """Merge two sorted lists into one sorted list."""
    result: List[T] = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])

    return result


def quick_sort(arr: List[T]) -> List[T]:
    """
    Quick Sort - divide and conquer with partitioning.

    Time Complexity: O(n log n) average, O(n²) worst (sorted input with bad pivot)
    Space Complexity: O(log n) average (recursion stack)
    Stable: No

    Generally fastest comparison sort in practice.
    Uses randomized pivot to avoid worst case.

    Args:
        arr: List to sort.

    Returns:
        New sorted list.
    """
    result = arr.copy()
    _quick_sort_helper(result, 0, len(result) - 1)
    return result


def _quick_sort_helper(arr: List[T], low: int, high: int) -> None:
    """Helper function for in-place quick sort."""
    if low < high:
        pivot_idx = _partition(arr, low, high)
        _quick_sort_helper(arr, low, pivot_idx - 1)
        _quick_sort_helper(arr, pivot_idx + 1, high)


def _partition(arr: List[T], low: int, high: int) -> int:
    """Partition array around a random pivot (Lomuto scheme)."""
    # Randomize pivot to avoid worst case
    pivot_idx = random.randint(low, high)
    arr[pivot_idx], arr[high] = arr[high], arr[pivot_idx]

    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def heap_sort(arr: List[T]) -> List[T]:
    """
    Heap Sort - build max-heap and extract elements.

    Time Complexity: O(n log n) all cases
    Space Complexity: O(1)
    Stable: No

    Guaranteed O(n log n) but typically slower than quicksort due to cache.

    Args:
        arr: List to sort.

    Returns:
        New sorted list.
    """
    result = arr.copy()
    n = len(result)

    # Build max-heap
    for i in range(n // 2 - 1, -1, -1):
        _heapify(result, n, i)

    # Extract elements
    for i in range(n - 1, 0, -1):
        result[0], result[i] = result[i], result[0]
        _heapify(result, i, 0)

    return result


def _heapify(arr: List[T], n: int, i: int) -> None:
    """Maintain max-heap property."""
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left

    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        _heapify(arr, n, largest)


def shell_sort(arr: List[T]) -> List[T]:
    """
    Shell Sort - generalization of insertion sort with gap sequence.

    Time Complexity: O(n²) worst, O(n log n) to O(n^1.3) average
    Space Complexity: O(1)
    Stable: No

    Uses Knuth's gap sequence: 1, 4, 13, 40, 121, ...

    Args:
        arr: List to sort.

    Returns:
        New sorted list.
    """
    result = arr.copy()
    n = len(result)

    # Generate Knuth's gap sequence
    gap = 1
    while gap < n // 3:
        gap = 3 * gap + 1

    while gap > 0:
        for i in range(gap, n):
            temp = result[i]
            j = i

            while j >= gap and result[j - gap] > temp:
                result[j] = result[j - gap]
                j -= gap

            result[j] = temp

        gap //= 3

    return result


# =============================================================================
# NON-COMPARISON SORTING ALGORITHMS
# =============================================================================


def counting_sort(arr: List[int]) -> List[int]:
    """
    Counting Sort - count occurrences, then reconstruct.

    Time Complexity: O(n + k) where k is range of values
    Space Complexity: O(k)
    Stable: Yes

    Only works for integers with known, reasonable range.

    Args:
        arr: List of integers to sort.

    Returns:
        New sorted list.
    """
    if not arr:
        return []

    min_val = min(arr)
    max_val = max(arr)
    range_size = max_val - min_val + 1

    # Count occurrences
    count = [0] * range_size
    for num in arr:
        count[num - min_val] += 1

    # Compute cumulative counts
    for i in range(1, range_size):
        count[i] += count[i - 1]

    # Build output array (iterate backwards for stability)
    result = [0] * len(arr)
    for num in reversed(arr):
        idx = count[num - min_val] - 1
        result[idx] = num
        count[num - min_val] -= 1

    return result


def radix_sort(arr: List[int]) -> List[int]:
    """
    Radix Sort - sort digit by digit using counting sort.

    Time Complexity: O(d * (n + k)) where d is digits, k is base
    Space Complexity: O(n + k)
    Stable: Yes

    Works for non-negative integers.

    Args:
        arr: List of non-negative integers to sort.

    Returns:
        New sorted list.
    """
    if not arr:
        return []

    # Handle negative numbers by separating them
    negatives = [-x for x in arr if x < 0]
    non_negatives = [x for x in arr if x >= 0]

    def radix_sort_positive(nums: List[int]) -> List[int]:
        if not nums:
            return []

        max_val = max(nums)
        result = nums.copy()
        exp = 1

        while max_val // exp > 0:
            result = _counting_sort_by_digit(result, exp)
            exp *= 10

        return result

    sorted_negatives = [-x for x in reversed(radix_sort_positive(negatives))]
    sorted_non_negatives = radix_sort_positive(non_negatives)

    return sorted_negatives + sorted_non_negatives


def _counting_sort_by_digit(arr: List[int], exp: int) -> List[int]:
    """Counting sort by a specific digit."""
    n = len(arr)
    result = [0] * n
    count = [0] * 10

    for num in arr:
        digit = (num // exp) % 10
        count[digit] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    for num in reversed(arr):
        digit = (num // exp) % 10
        idx = count[digit] - 1
        result[idx] = num
        count[digit] -= 1

    return result


def bucket_sort(arr: List[float], num_buckets: int = 10) -> List[float]:
    """
    Bucket Sort - distribute into buckets, sort each, concatenate.

    Time Complexity: O(n + k) average, O(n²) worst
    Space Complexity: O(n)
    Stable: Yes (if using stable sort for buckets)

    Best for uniformly distributed floating-point numbers in [0, 1).

    Args:
        arr: List of floats in [0, 1) to sort.
        num_buckets: Number of buckets.

    Returns:
        New sorted list.
    """
    if not arr:
        return []

    # Create buckets
    buckets: List[List[float]] = [[] for _ in range(num_buckets)]

    # Find min and max for scaling
    min_val = min(arr)
    max_val = max(arr)
    range_val = max_val - min_val

    if range_val == 0:
        return arr.copy()

    # Distribute elements into buckets
    for num in arr:
        # Scale to [0, num_buckets)
        idx = int((num - min_val) / range_val * (num_buckets - 1))
        buckets[idx].append(num)

    # Sort each bucket using insertion sort
    result: List[float] = []
    for bucket in buckets:
        # Use insertion sort for small buckets
        bucket.sort()
        result.extend(bucket)

    return result


# =============================================================================
# QUICK SELECT (Related algorithm)
# =============================================================================


def quick_select(arr: List[T], k: int) -> T:
    """
    Quick Select - find kth smallest element.

    Time Complexity: O(n) average, O(n²) worst
    Space Complexity: O(1)

    Uses randomized partitioning.

    Args:
        arr: List to search.
        k: Position (1-indexed, k=1 means smallest).

    Returns:
        The kth smallest element.

    Raises:
        ValueError: If k is out of range.
    """
    if k < 1 or k > len(arr):
        raise ValueError(f"k={k} out of range for array of size {len(arr)}")

    result = arr.copy()
    return _quick_select_helper(result, 0, len(result) - 1, k - 1)


def _quick_select_helper(arr: List[T], low: int, high: int, k: int) -> T:
    """Helper for quick select."""
    if low == high:
        return arr[low]

    pivot_idx = _partition(arr, low, high)

    if k == pivot_idx:
        return arr[k]
    elif k < pivot_idx:
        return _quick_select_helper(arr, low, pivot_idx - 1, k)
    else:
        return _quick_select_helper(arr, pivot_idx + 1, high, k)


# =============================================================================
# SORTING WITH CUSTOM COMPARATOR
# =============================================================================


def sort_with_comparator(
    arr: List[T],
    comparator: Callable[[T, T], int]
) -> List[T]:
    """
    Sort using a custom comparator function.

    Uses merge sort for stability.

    Args:
        arr: List to sort.
        comparator: Function that returns:
                   - negative if a < b
                   - zero if a == b
                   - positive if a > b

    Returns:
        New sorted list.
    """
    if len(arr) <= 1:
        return arr.copy()

    mid = len(arr) // 2
    left = sort_with_comparator(arr[:mid], comparator)
    right = sort_with_comparator(arr[mid:], comparator)

    return _merge_with_comparator(left, right, comparator)


def _merge_with_comparator(
    left: List[T],
    right: List[T],
    comparator: Callable[[T, T], int]
) -> List[T]:
    """Merge with custom comparator."""
    result: List[T] = []
    i = j = 0

    while i < len(left) and j < len(right):
        if comparator(left[i], right[j]) <= 0:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])

    return result


# =============================================================================
# UTILITY FUNCTIONS
# =============================================================================


def is_sorted(arr: List[T], reverse: bool = False) -> bool:
    """
    Check if array is sorted.

    Args:
        arr: List to check.
        reverse: If True, check for descending order.

    Returns:
        True if sorted.
    """
    if len(arr) <= 1:
        return True

    if reverse:
        return all(arr[i] >= arr[i + 1] for i in range(len(arr) - 1))
    else:
        return all(arr[i] <= arr[i + 1] for i in range(len(arr) - 1))

File: by-language/c/algorithms/sorting.c
/**
 * Sorting Algorithms Implementation
 * See sorting.h for documentation
 */

#include "sorting.h"
#include <stdlib.h>
#include <string.h>
#include <limits.h>

/* ============== Helper Functions ============== */

static void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

/* ============== Bubble Sort ============== */

void bubble_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    for (size_t i = 0; i < n - 1; i++) {
        bool swapped = false;
        for (size_t j = 0; j < n - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(&arr[j], &arr[j + 1]);
                swapped = true;
            }
        }
        if (!swapped) {
            break; /* Already sorted */
        }
    }
}

/* ============== Selection Sort ============== */

void selection_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    for (size_t i = 0; i < n - 1; i++) {
        size_t min_idx = i;
        for (size_t j = i + 1; j < n; j++) {
            if (arr[j] < arr[min_idx]) {
                min_idx = j;
            }
        }
        if (min_idx != i) {
            swap(&arr[i], &arr[min_idx]);
        }
    }
}

/* ============== Insertion Sort ============== */

void insertion_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    for (size_t i = 1; i < n; i++) {
        int key = arr[i];
        size_t j = i;
        while (j > 0 && arr[j - 1] > key) {
            arr[j] = arr[j - 1];
            j--;
        }
        arr[j] = key;
    }
}

/* ============== Merge Sort ============== */

static void merge(int *arr, int *temp, size_t left, size_t mid, size_t right) {
    size_t i = left;
    size_t j = mid + 1;
    size_t k = left;

    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }

    while (i <= mid) {
        temp[k++] = arr[i++];
    }

    while (j <= right) {
        temp[k++] = arr[j++];
    }

    for (i = left; i <= right; i++) {
        arr[i] = temp[i];
    }
}

static void merge_sort_recursive(int *arr, int *temp, size_t left, size_t right) {
    if (left < right) {
        size_t mid = left + (right - left) / 2;
        merge_sort_recursive(arr, temp, left, mid);
        merge_sort_recursive(arr, temp, mid + 1, right);
        merge(arr, temp, left, mid, right);
    }
}

void merge_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    int *temp = malloc(n * sizeof(int));
    if (temp == NULL) {
        return;
    }

    merge_sort_recursive(arr, temp, 0, n - 1);
    free(temp);
}

/* ============== Quick Sort ============== */

static size_t partition(int *arr, size_t low, size_t high) {
    int pivot = arr[high];
    size_t i = low;

    for (size_t j = low; j < high; j++) {
        if (arr[j] < pivot) {
            swap(&arr[i], &arr[j]);
            i++;
        }
    }
    swap(&arr[i], &arr[high]);
    return i;
}

static void quick_sort_recursive(int *arr, size_t low, size_t high) {
    if (low < high) {
        size_t pi = partition(arr, low, high);
        if (pi > 0) {
            quick_sort_recursive(arr, low, pi - 1);
        }
        quick_sort_recursive(arr, pi + 1, high);
    }
}

void quick_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }
    quick_sort_recursive(arr, 0, n - 1);
}

/* ============== Quick Sort with Median-of-Three ============== */

static size_t median_of_three(int *arr, size_t low, size_t high) {
    size_t mid = low + (high - low) / 2;

    if (arr[low] > arr[mid]) {
        swap(&arr[low], &arr[mid]);
    }
    if (arr[low] > arr[high]) {
        swap(&arr[low], &arr[high]);
    }
    if (arr[mid] > arr[high]) {
        swap(&arr[mid], &arr[high]);
    }

    swap(&arr[mid], &arr[high - 1]);
    return high - 1;
}

static size_t partition_median(int *arr, size_t low, size_t high) {
    if (high - low < 3) {
        return partition(arr, low, high);
    }

    size_t pivot_idx = median_of_three(arr, low, high);
    int pivot = arr[pivot_idx];
    size_t i = low;
    size_t j = high - 1;

    while (true) {
        while (arr[++i] < pivot);
        while (arr[--j] > pivot);
        if (i >= j) {
            break;
        }
        swap(&arr[i], &arr[j]);
    }
    swap(&arr[i], &arr[high - 1]);
    return i;
}

static void quick_sort_median_recursive(int *arr, size_t low, size_t high) {
    if (low + 10 > high) {
        /* Use insertion sort for small subarrays */
        for (size_t i = low + 1; i <= high; i++) {
            int key = arr[i];
            size_t j = i;
            while (j > low && arr[j - 1] > key) {
                arr[j] = arr[j - 1];
                j--;
            }
            arr[j] = key;
        }
        return;
    }

    size_t pi = partition_median(arr, low, high);
    if (pi > low) {
        quick_sort_median_recursive(arr, low, pi - 1);
    }
    quick_sort_median_recursive(arr, pi + 1, high);
}

void quick_sort_median(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }
    quick_sort_median_recursive(arr, 0, n - 1);
}

/* ============== Counting Sort ============== */

void counting_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    /* Find range */
    int min_val = arr[0];
    int max_val = arr[0];
    for (size_t i = 1; i < n; i++) {
        if (arr[i] < min_val) min_val = arr[i];
        if (arr[i] > max_val) max_val = arr[i];
    }

    size_t range = (size_t)(max_val - min_val + 1);
    int *count = calloc(range, sizeof(int));
    int *output = malloc(n * sizeof(int));

    if (count == NULL || output == NULL) {
        free(count);
        free(output);
        return;
    }

    /* Count occurrences */
    for (size_t i = 0; i < n; i++) {
        count[arr[i] - min_val]++;
    }

    /* Cumulative count */
    for (size_t i = 1; i < range; i++) {
        count[i] += count[i - 1];
    }

    /* Build output (stable) */
    for (size_t i = n; i > 0; i--) {
        output[count[arr[i - 1] - min_val] - 1] = arr[i - 1];
        count[arr[i - 1] - min_val]--;
    }

    /* Copy back */
    memcpy(arr, output, n * sizeof(int));

    free(count);
    free(output);
}

/* ============== Radix Sort (LSD) ============== */

static void counting_sort_digit(int *arr, size_t n, int exp) {
    int *output = malloc(n * sizeof(int));
    int count[10] = {0};

    if (output == NULL) {
        return;
    }

    /* Count occurrences of each digit */
    for (size_t i = 0; i < n; i++) {
        count[(arr[i] / exp) % 10]++;
    }

    /* Cumulative count */
    for (int i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    /* Build output (stable) */
    for (size_t i = n; i > 0; i--) {
        int digit = (arr[i - 1] / exp) % 10;
        output[count[digit] - 1] = arr[i - 1];
        count[digit]--;
    }

    /* Copy back */
    memcpy(arr, output, n * sizeof(int));
    free(output);
}

void radix_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    /* Find max to know number of digits */
    int max_val = arr[0];
    for (size_t i = 1; i < n; i++) {
        if (arr[i] > max_val) {
            max_val = arr[i];
        }
    }

    /* Sort by each digit */
    for (int exp = 1; max_val / exp > 0; exp *= 10) {
        counting_sort_digit(arr, n, exp);
    }
}

/* ============== Shell Sort ============== */

void shell_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    /* Marcin Ciura's gap sequence */
    size_t gaps[] = {701, 301, 132, 57, 23, 10, 4, 1};
    size_t num_gaps = sizeof(gaps) / sizeof(gaps[0]);

    for (size_t g = 0; g < num_gaps; g++) {
        size_t gap = gaps[g];
        if (gap >= n) continue;

        for (size_t i = gap; i < n; i++) {
            int temp = arr[i];
            size_t j = i;
            while (j >= gap && arr[j - gap] > temp) {
                arr[j] = arr[j - gap];
                j -= gap;
            }
            arr[j] = temp;
        }
    }
}

/* ============== Dutch National Flag ============== */

void dutch_flag_sort(int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return;
    }

    size_t low = 0;
    size_t mid = 0;
    size_t high = n - 1;

    while (mid <= high) {
        if (arr[mid] == 0) {
            swap(&arr[low], &arr[mid]);
            low++;
            mid++;
        } else if (arr[mid] == 1) {
            mid++;
        } else { /* arr[mid] == 2 */
            swap(&arr[mid], &arr[high]);
            if (high == 0) break;
            high--;
        }
    }
}

/* ============== Quick Select ============== */

static size_t partition_select(int *arr, size_t low, size_t high) {
    int pivot = arr[high];
    size_t i = low;

    for (size_t j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            swap(&arr[i], &arr[j]);
            i++;
        }
    }
    swap(&arr[i], &arr[high]);
    return i;
}

int quick_select(int *arr, size_t n, size_t k) {
    if (arr == NULL || n == 0 || k >= n) {
        return INT_MIN;
    }

    /* Make a copy to avoid modifying original */
    int *copy = malloc(n * sizeof(int));
    if (copy == NULL) {
        return INT_MIN;
    }
    memcpy(copy, arr, n * sizeof(int));

    size_t low = 0;
    size_t high = n - 1;

    while (low <= high) {
        size_t pi = partition_select(copy, low, high);

        if (pi == k) {
            int result = copy[pi];
            free(copy);
            return result;
        } else if (pi < k) {
            low = pi + 1;
        } else {
            if (pi == 0) break;
            high = pi - 1;
        }
    }

    free(copy);
    return INT_MIN;
}

/* ============== Utility Functions ============== */

bool is_sorted(const int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return true;
    }

    for (size_t i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

bool is_sorted_desc(const int *arr, size_t n) {
    if (arr == NULL || n <= 1) {
        return true;
    }

    for (size_t i = 0; i < n - 1; i++) {
        if (arr[i] < arr[i + 1]) {
            return false;
        }
    }
    return true;
}
